### DISCUSSION TITLE
How to mock in tests?

### DISCUSSION THREAD

**Discussion by lumos42** [aptos-labs/aptos-developer-discussions#349] [General]
Hi, Is there a way to mock function calls to write unit tests?

Let's say module A calls module B, and I want to test module A only. Is there a way to mock calls to module B such that tests can be isolated?

**Comment by 0xbrw** [comment]
You can look through this code example, module A would be aptogotchi.move and module B would be food.move: 
https://github.com/aptos-labs/aptogotchi-intermediate/blob/main/move/sources/aptogotchi.move

As for mocking you can create a `#test_only` function to help perform your mock.

**Reply by 0xbrw** [reply]
Another option is in your Move.toml file you can specify module B as a `[dev-dependencies]` and mock the desired function in module B. Just be sure to run test with the `--dev` flag. For example, `aptos move test --dev`

**Reply by lumos42** [reply]
Perhaps I didn't explain my question well. I'm looking to mock function calls, so in your example, let's say I wanted to test the `aptogotchi::feed` function, but the function calls `food::burn_food`. Now my unit tests for `aptogotchi::feed` should not depend on other modules, so is there a way to mock calls to `food::burn_food` such that while testing it doesn't make an actual call to food package but uses the mock response provided by me? This is a common functionality provided in languages for ease of testing.

More details [here](https://en.wikipedia.org/wiki/Mock_object#Motivation).

The way to do it in Solidity is to have an interface IFood and have an implementation for a real food module and one for mock food. And then for test setup we can just configure the aptogotchi contract to use mock food.

**Reply by 0xbrw** [reply]
Aptos does not have that mocking functionality yet. For now if you don't want to make a call into module B, I would create a test only helper function.
